File: leetCode/easy/88-mergeSortedArr.ts
import { assertEquals } from "https://deno.land/std@0.182.0/testing/asserts.ts";
import {
  describe,
  it,
  beforeEach,
  afterEach,
} from "https://deno.land/std@0.182.0/testing/bdd.ts";

function mergeSortedArr(
  nums1: number[],
  m: number,
  nums2: number[],
  n: number
) {
  let left = m - 1;
  let right = n - 1;
  const numsLen = nums1.length;

  for (let i = numsLen - 1; i >= 0; i -= 1) {
    if (right < 0) break;

    if (left >= 0 && nums1[left] > nums2[right]) {
      nums1[i] = nums1[left];
      left -= 1;
    } else {
      nums1[i] = nums2[right];
      right -= 1;
    }
  }
}

export { mergeSortedArr };

describe("mergeSortedArr", () => {
  it("1. should merge two sorted arrays", () => {
    const nums1 = [1, 2, 3, 0, 0, 0];
    const m = 3;
    const nums2 = [2, 5, 6];
    const n = 3;
    const output = [1, 2, 2, 3, 5, 6];
    mergeSortedArr(nums1, m, nums2, n);
    assertEquals(nums1, output);
  });

  it("2. should merge two sorted arrays", () => {
    const nums1 = [1];
    const m = 1;
    const nums2: number[] = [];
    const n = 0;
    const output = [1];
    mergeSortedArr(nums1, m, nums2, n);
    assertEquals(nums1, output);
  });

  it("3. should merge two sorted arrays", () => {
    const nums1 = [0];
    const m = 0;
    const nums2 = [1];
    const n = 1;
    const output = [1];
    mergeSortedArr(nums1, m, nums2, n);
    assertEquals(nums1, output);
  });

  it("4. should merge two sorted arrays", () => {
    const nums1 = [4, 5, 6, 0, 0, 0];
    const m = 3;
    const nums2 = [1, 2, 3];
    const n = 3;
    const output = [1, 2, 3, 4, 5, 6];
    mergeSortedArr(nums1, m, nums2, n);
    assertEquals(nums1, output);
  });
});

File: leetCode/easy/88-mergeSortedArr.jl
function mergeSortedArr(nums1::Array{T,1}, m::T, nums2::Array{T,1}, n::T) where {T<:Integer}
  left = m
  right = n
  numsLen = length(nums1)

  for i in reverse(1:numsLen)
    if right < 1
      break
    end

    if left >= 1 && nums1[left] > nums2[right]
      nums1[i] = nums1[left]
      left -= 1
    else
      nums1[i] = nums2[right]
      right -= 1
    end
  end
end

export mergeSortedArr

using Test

Test.@testset "mergeSortedArr" begin
  nums1 = [1, 2, 3, 0, 0, 0]
  nums2 = [2, 5, 6]
  mergeSortedArr(nums1, 3, nums2, 3)
  Test.@test nums1 == [1, 2, 2, 3, 5, 6]

  nums1 = [1]
  nums2::Array{Int64,1} = []
  mergeSortedArr(nums1, 1, nums2, 0)
  Test.@test nums1 == [1]

  nums1 = [0]
  nums2 = [1]
  mergeSortedArr(nums1, 0, nums2, 1)
  Test.@test nums1 == [1]

  nums1 = [4, 0, 0, 0, 0, 0]
  nums2 = [1, 2, 3, 5, 6]
  mergeSortedArr(nums1, 1, nums2, 5)
  Test.@test nums1 == [1, 2, 3, 4, 5, 6]
end